crates/flowistry_pdg_construction/src/graph.rs

```diff
@@ -189,3 +232,201 @@ impl<'tcx> DepGraph<'tcx> {
+    fn encode(&self, e: &mut E) {
+        let s: &String = &self.0;
+        s.encode(e);
+    }
+}
+
+impl<D: Decoder> Decodable<D> for InternedString {
+    fn decode(d: &mut D) -> Self {
+        Self(Intern::new(String::decode(d)))
+    }
+}
+
+/// A PDG that is fit for combining with other PDGs
+#[derive(Debug, Clone, TyDecodable, TyEncodable)]
+pub struct PartialGraph<'tcx> {
+    pub(crate) nodes: FxHashSet<DepNode<'tcx>>,
+    pub(crate) edges: FxHashSet<(DepNode<'tcx>, DepNode<'tcx>, DepEdge)>,
+    pub(crate) monos: FxHashMap<CallString, GenericArgsRef<'tcx>>,
+    pub(crate) generics: GenericArgsRef<'tcx>,
+    pub(crate) asyncness: Asyncness,
+    def_id: DefId,
+    arg_count: usize,
+}
+
+impl<'tcx> PartialGraph<'tcx> {
+    pub fn mentioned_call_string<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = CallString> + Captures<'tcx> + 'a {
+        self.nodes
+            .iter()
+            .map(|n| &n.at)
+            .chain(self.edges.iter().map(|e| &e.2.at))
+            .copied()
+    }
+
+    pub fn get_mono(&self, cs: CallString) -> Option<GenericArgsRef<'tcx>> {
+        if let Some(caller) = cs.caller() {
+            self.monos.get(&caller).copied()
+        } else {
+            Some(self.generics)
+        }
+    }
+
+    pub fn asyncness(&self) -> Asyncness {
+        self.asyncness
+    }
+
+    pub fn new(
+        asyncness: Asyncness,
+        generics: GenericArgsRef<'tcx>,
+        def_id: DefId,
+        arg_count: usize,
+    ) -> Self {
+        Self {
+            nodes: Default::default(),
+            edges: Default::default(),
+            monos: Default::default(),
+            generics,
+            asyncness,
+            def_id,
+            arg_count,
+        }
+    }
+
+    pub(crate) fn parentable_srcs<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = (DepNode<'tcx>, Option<u8>)> + 'a {
+        self.edges
+            .iter()
+            .map(|(src, _, _)| *src)
+            .filter_map(move |a| Some((a, as_arg(&a, self.def_id, self.arg_count)?)))
+            .filter(|(node, _)| node.at.leaf().location.is_start())
+    }
+
+    pub(crate) fn parentable_dsts<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = (DepNode<'tcx>, Option<u8>)> + 'a {
+        self.edges
+            .iter()
+            .map(|(_, dst, _)| *dst)
+            .filter_map(move |a| Some((a, as_arg(&a, self.def_id, self.arg_count)?)))
+            .filter(|node| node.0.at.leaf().location.is_end())
+    }
+}
+
+fn as_arg(node: &DepNode<'_>, def_id: DefId, arg_count: usize) -> Option<Option<u8>> {
+    if node.at.leaf().function != def_id {
+        return None;
+    }
+    let local = node.place.local.as_usize();
+    if node.place.local == rustc_middle::mir::RETURN_PLACE {
+        Some(None)
+    } else if local > 0 && (local - 1) < arg_count {
+        Some(Some(node.place.local.as_u32() as u8 - 1))
+    } else {
+        None
+    }
+}
+
+impl<'tcx> TransformCallString for PartialGraph<'tcx> {
```

Comment:
Is that necessary? Visitors are also less extensible because you need a visit method for each type you want to be able to traverse. That feels like overengineering in this case. I can make the function `impl FnMut(&mut CallString)` if you like that makes it a bit more general I guess